vmss-prototype/vmss-prototype

```diff
@@ -1014,32 +1014,41 @@ def vmss_prototype_update(sub_args):
             fatal_error('VMSS {0} failed to be configured to use prototype iamge'.format(vmss))
 
     # Get the VMSS extensions and see if we can clean them up
-    output, _, _ = run(az(['vmss', 'extension', 'list'], subscription) + [
-                       '--resource-group', resource_group,
-                       '--vmss-name', vmss
-                       ], retries=3)
-    extensions = json.loads(output)
-
-    for extension in extensions:
-        # Now that we have a disk image that already has all of our
-        # provisioning bits on it (prototype) we can remove all extensions
-        # But...  We want to keep the AKS Engine-identifying extension for telemetry reasons
-        # It is a no-op code wise but gives counts
-        if 'vmss-computeAksLinuxBilling' not in extension['name']:
-            run(az(['vmss', 'extension', 'delete'], subscription) + [
-                '--resource-group', resource_group,
-                '--vmss-name', vmss,
-                '--name', extension['name']
-                ], retries=3, check=False, retry_func=not_found_no_retry)
+    # Note that we could potentially loop here since we have seen a delete
+    # report success but not actually delete the extension.  So we can try
+    # this loop up to 5 times (just in case)
+    delete_tries = 5
+    delete_extensions = True
+    while delete_extensions and delete_tries > 0:
+        delete_tries = delete_tries - 1
+        delete_extensions = False
+        output, _, _ = run(az(['vmss', 'extension', 'list'], subscription) + [
+                           '--resource-group', resource_group,
+                           '--vmss-name', vmss
+                           ], retries=3)
+        extensions = json.loads(output)
+
+        for extension in extensions:
+            # Now that we have a disk image that already has all of our
+            # provisioning bits on it (prototype) we can remove all extensions
+            # But...  We want to keep the AKS Engine-identifying extension for telemetry reasons
+            # It is a no-op code wise but gives counts
+            if 'vmss-computeAksLinuxBilling' not in extension['name']:
+                delete_extensions = True
+                run(az(['vmss', 'extension', 'delete'], subscription) + [
+                    '--resource-group', resource_group,
+                    '--vmss-name', vmss,
+                    '--name', extension['name']
+                    ], retries=3, check=False, retry_func=not_found_no_retry, log_retry_errors=True)
```

Comment:
Should we add a sleep at the end of this while loop? In the event that our follow-up `vmss extension list` operation still returns an extension (not the `vmss-computeAksLinuxBilling` one), our 3 remaining retries will happen in quick succession.

Maybe we should add a simple thing at the top of this loop:

```python
if delete_tries < 4:
    sleep 30 # caveat emptor I don't know the correct python sleep statement off the top of my head :)
```

Reply:
Not a bad idea - to sleep based on how much we have done.  Let me look into it.

Reply:
However, the tricky bit is that I really don't want to sleep if we have correctly deleted everything.  We always loop twice, the second time just notices that all is deleted.

Reply:
Note that the backoff is likely redundant since it takes seconds (like 10+ seconds) when an extension is deleted (as per the logs I captured) so it was already somewhat slow roll already.

However, it can not hurt to put the backoff logic in to help.